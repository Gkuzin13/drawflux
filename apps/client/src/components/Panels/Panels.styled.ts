import { styled } from 'shared';
import * as ButtonStyled from '@/components/Elements/Button/Button.styled';

export const Container = styled('div', {
  position: 'fixed',
  display: 'flex',
  justifyContent: 'space-between',
  flexDirection: 'column',
  inset: 0,
  width: '100%',
  height: '100%',
  pointerEvents: 'none',
  zIndex: 1,
  padding: '$3',
});

export const Panel = styled('div', {
  display: 'flex',
  justifyContent: 'space-between',
  gap: '$1',
  padding: '$1',
  pointerEvents: 'all',
  boxShadow: '$small',
  borderRadius: '$1',
  backgroundColor: '$white50',
});

export const Button = styled(ButtonStyled.Button, {
  defaultVariants: {
    size: 'extra-small',
    color: 'secondary',
    squared: true,
  },
});

export const TopPanel = styled('div', {
  position: 'relative',
  display: 'flex',
  justifyContent: 'space-between',
  width: '100%',
  gap: '$2',
});

export const TopPanelRightContainer = styled(Panel, {});

export const BottomPanel = styled('div', {
  display: 'flex',
  justifyContent: 'center',
});
