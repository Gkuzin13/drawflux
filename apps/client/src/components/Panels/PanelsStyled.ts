import { styled } from 'shared';
import { ButtonStyled } from '../core/Button/ButtonStyled';

export const PanelsContainer = styled('div', {
  position: 'fixed',
  display: 'flex',
  justifyContent: 'space-between',
  flexDirection: 'column',
  inset: 0,
  width: '100%',
  height: '100%',
  pointerEvents: 'none',
  zIndex: 1,
  padding: '$3',
});

export const Panel = styled('div', {
  display: 'flex',
  justifyContent: 'space-between',
  gap: '$1',
  padding: '$1',
  pointerEvents: 'all',
  boxShadow: '$small',
  borderRadius: '$1',
  backgroundColor: '$white50',
});

export const PanelButton = styled(ButtonStyled, {
  defaultVariants: {
    size: 'extra-small',
    color: 'secondary',
    squared: true,
  },
});

export const TopPanel = styled('div', {
  position: 'relative',
  display: 'flex',
  justifyContent: 'space-between',
  width: '100%',
  gap: '$2',
});

export const TopPanelRightContainer = styled(Panel, {});

export const BottomPanel = styled('div', {
  display: 'flex',
  justifyContent: 'center',
});
