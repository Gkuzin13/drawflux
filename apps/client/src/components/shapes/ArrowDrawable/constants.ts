export const ANCHOR_INDEX = {
  START: 0,
  END: 1,
  CONTROL: 2,
} as const;

export const DEFAULT_BEND = 0.5;
