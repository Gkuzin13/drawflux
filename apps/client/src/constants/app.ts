import { Schemas } from 'shared';
import { z } from 'zod';

export const BASE_URL = 'https://drawflux-api.onrender.com';
export const BASE_URL_DEV = 'http://localhost:7456';

export const BASE_WS_URL = 'wss://drawflux-api.onrender.com';
export const BASE_WS_URL_DEV = 'ws://localhost:7456';

export const IS_PROD = process.env.NODE_ENV === 'production';

export const PAGE_URL_SEARCH_PARAM_KEY = 'page';
export const LOCAL_STORAGE_KEY = 'drawflux';

export const WS_THROTTLE_MS = 16;

export const PROJECT_FILE_EXT = 'drawflux';
export const PROJECT_FILE_NAME = 'project';

export const USER = {
  maxNameLength: 10,
};

export const ZOOM_RANGE = {
  MIN: 0.1,
  MAX: 2.5,
};
export const ZOOM_MULTIPLIER_VALUE = 0.1;
export const DEFAULT_ZOOM_VALUE = 1;

export const appState = z.object({
  page: z.object({
    ...Schemas.Page.shape.page.shape,
    toolType: z.union([
      ...Schemas.Node.shape.type.options,
      z.literal('hand'),
      z.literal('select'),
    ]),
    selectedNodesIds: z.record(z.string(), z.boolean()),
  }),
});

export type AppState = z.infer<typeof appState>;
