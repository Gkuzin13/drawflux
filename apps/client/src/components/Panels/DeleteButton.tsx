import Icon from '../Elements/Icon/Icon';
import { DELETE_NODES } from '@/constants/panels';
import { createKeyTitle } from '@/utils/string';
import { memo } from 'react';
import * as Styled from './Panels.styled';

type Props = {
  disabled: boolean;
  onClick: () => void;
};

const DeleteButton = ({ disabled, onClick }: Props) => {
  return (
    <Styled.Button
      title={createKeyTitle(DELETE_NODES.name, [
        ...DELETE_NODES.modifierKeys,
        DELETE_NODES.key,
      ])}
      disabled={disabled}
      onClick={onClick}
    >
      <Icon name={DELETE_NODES.icon} />
    </Styled.Button>
  );
};

export default memo(DeleteButton);
