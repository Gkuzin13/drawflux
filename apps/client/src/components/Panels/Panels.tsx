import type Konva from 'konva';
import { type RefObject, useMemo } from 'react';
import { type NodeStyle, type NodeObject, Schemas } from 'shared';
import { z } from 'zod';
import { type MenuPanelActionType } from '@/constants/menu';
import { type Tool } from '@/constants/tool';
import { useAppDispatch, useAppSelector } from '@/stores/hooks';
import { controlActions, selectControl } from '@/stores/slices/controlSlice';
import { modalActions } from '@/stores/slices/modalSlice';
import { nodesActions, selectNodes } from '@/stores/slices/nodesSlice';
import {
  selectStageConfig,
  stageConfigActions,
  type StageConfigState,
} from '@/stores/slices/stageConfigSlice';
import { store } from '@/stores/store';
import { downloadDataUrlAsFile, loadJsonFile } from '@/utils/file';
import ControlPanel from './ControlPanel/ControlPanel';
import MenuPanel from './MenuPanel/MenuPanel';
import SharePanel from './SharePanel/SharePanel';
import StylePanel, { type StylePanelProps } from './StylePanel/StylePanel';
import ToolsDock from './ToolsPanel/ToolsPanel';
import ZoomPanel from './ZoomPanel/ZoomPanel';

type Props = {
  stageRef: RefObject<Konva.Stage>;
  isPageShared?: boolean;
};

const Panels = ({ stageRef, isPageShared = false }: Props) => {
  const { selectedNodeId, toolType } = useAppSelector(selectControl);
  const stageConfig = useAppSelector(selectStageConfig);

  const { past, present, future } = useAppSelector(selectNodes);

  const nodes = present.nodes;

  const dispatch = useAppDispatch();

  const selectedNode = useMemo(() => {
    return nodes.find((n) => n.nodeProps.id === selectedNodeId);
  }, [selectedNodeId, nodes]);

  const stylePanelEnabledOptions = useMemo<
    StylePanelProps['enabledOptions']
  >(() => {
    switch (selectedNode?.type) {
      case 'text':
        return { line: false, size: true };
      default:
        return { line: true, size: true };
    }
  }, [selectedNode]);

  const onToolTypeChange = (type: Tool['value']) => {
    dispatch(controlActions.setToolType(type));
  };

  const handleStyleChange = (style: NodeStyle) => {
    if (!selectedNode) return;

    const updatedNode = { ...selectedNode, style };

    dispatch(nodesActions.update([updatedNode]));
  };

  const handleMenuAction = (type: MenuPanelActionType) => {
    switch (type) {
      case 'export-as-image': {
        const dataUrl = stageRef.current?.toDataURL();

        if (dataUrl) {
          downloadDataUrlAsFile(dataUrl, 'export-image');
        }
        break;
      }
      case 'export-as-json': {
        const state = store.getState();

        const stateToExport = {
          stageConfig: state.stageConfig,
          nodes: state.nodesHistory.present.nodes,
        };

        const dataUrl = URL.createObjectURL(
          new Blob([JSON.stringify(stateToExport)], {
            type: 'application/json',
          }),
        );

        downloadDataUrlAsFile(dataUrl, 'export-json');
        break;
      }
      case 'import-json': {
        const schema = z.object({
          nodes: Schemas.Node.array(),
          stageConfig: z.unknown(),
        });

        loadJsonFile<{ nodes: NodeObject[]; stageConfig: StageConfigState }>(
          schema,
        ).then((state) => {
          if (!state) {
            dispatch(
              modalActions.open({
                title: 'Error',
                message: 'Could not load file',
              }),
            );
            return;
          }
          dispatch(nodesActions.set(state.nodes));
          dispatch(stageConfigActions.set(state.stageConfig));
        });
        break;
      }
    }
  };

  const handleZoomChange = (value: number) => {
    dispatch(stageConfigActions.set({ ...stageConfig, scale: value }));
  };

  return (
    <>
      <SharePanel
        isPageShared={isPageShared}
        pageState={{ page: { nodes, stageConfig } }}
      />
      <MenuPanel onAction={handleMenuAction} />
      <ToolsDock activeTool={toolType} onToolSelect={onToolTypeChange} />
      {selectedNode && (
        <StylePanel
          style={selectedNode.style}
          enabledOptions={stylePanelEnabledOptions}
          onStyleChange={handleStyleChange}
        />
      )}
      <ControlPanel
        onControl={dispatch}
        enabledControls={{
          undo: Boolean(past.length),
          redo: Boolean(future.length),
          clear: Boolean(nodes.length),
        }}
      />
      <ZoomPanel value={stageConfig.scale} onZoomChange={handleZoomChange} />
    </>
  );
};

export default Panels;
