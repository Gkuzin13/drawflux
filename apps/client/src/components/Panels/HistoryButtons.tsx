import { memo } from 'react';
import Icon from '../Elements/Icon/Icon';
import { HISTORY } from '@/constants/panels';
import { createKeyTitle } from '@/utils/string';
import * as Styled from './Panels.styled';
import type { HistoryControlKey } from '@/constants/panels';

type Props = {
  disabledUndo: boolean;
  disabledRedo: boolean;
  onClick: (type: HistoryControlKey) => void;
};

const { undo, redo } = HISTORY;

const HistoryButtons = ({ disabledUndo, disabledRedo, onClick }: Props) => {
  return (
    <>
      <Styled.Button
        title={createKeyTitle(undo.name, [...undo.modifierKeys, undo.key])}
        disabled={disabledUndo}
        onClick={() => onClick(undo.value)}
        data-testid="undo-history-button"
      >
        <Icon name={undo.icon} />
      </Styled.Button>
      <Styled.Button
        title={createKeyTitle(redo.name, [...redo.modifierKeys, redo.key])}
        disabled={disabledRedo}
        onClick={() => onClick(redo.value)}
        data-testid="redo-history-button"
      >
        <Icon name={redo.icon} />
      </Styled.Button>
    </>
  );
};

export default memo(HistoryButtons);
