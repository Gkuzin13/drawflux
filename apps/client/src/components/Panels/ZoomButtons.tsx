import { memo } from 'react';
import { DEFAULT_ZOOM_VALUE, ZOOM_RANGE } from '@/constants/app';
import { ZOOM, type ZoomActionKey } from '@/constants/panels';
import Icon from '@/components/Elements/Icon/Icon';
import { createKeyTitle } from '@/utils/string';
import * as Styled from './Panels.styled';

type Props = {
  value: number;
  onZoomChange: (action: ZoomActionKey) => void;
};

const ZoomPanel = ({ value, onZoomChange }: Props) => {
  const stageScalePercent = `${Math.round(value * 100)}%`;

  const handleZoomAction = (action: ZoomActionKey) => {
    onZoomChange(action);
  };

  return (
    <>
      <Styled.Button
        disabled={value === DEFAULT_ZOOM_VALUE}
        title={ZOOM.reset.name}
        data-testid="zoom-reset-button"
        css={{ fontSize: '$1', width: 'calc($5 * 1.5)' }}
        onClick={() => handleZoomAction('reset')}
      >
        {stageScalePercent}
      </Styled.Button>
      <Styled.Button
        disabled={value === ZOOM_RANGE[1]}
        title={createKeyTitle(ZOOM.in.name, [
          ZOOM.in.key,
          ...ZOOM.in.modifierKeys,
        ])}
        data-testid="zoom-in-button"
        onClick={() => handleZoomAction('in')}
      >
        <Icon name={ZOOM.in.icon} size="sm" />
      </Styled.Button>
      <Styled.Button
        disabled={value === ZOOM_RANGE[0]}
        title={createKeyTitle(ZOOM.out.name, [
          ZOOM.out.key,
          ...ZOOM.out.modifierKeys,
        ])}
        data-testid="zoom-out-button"
        onClick={() => handleZoomAction('out')}
      >
        <Icon name={ZOOM.out.icon} size="sm" />
      </Styled.Button>
    </>
  );
};

export default memo(ZoomPanel);
