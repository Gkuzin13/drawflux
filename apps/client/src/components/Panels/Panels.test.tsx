import { screen, within } from '@testing-library/react';
import { nodesGenerator, stateGenerator } from '@/test/data-generators';
import { renderWithProviders } from '@/test/test-utils';
import Panels from './Panels';
import {
  canvasActions,
  initialState as initialCanvasState,
} from '@/services/canvas/slice';
import { historyActions } from '@/stores/reducers/history';
import { DEFAULT_ZOOM_VALUE, ZOOM_RANGE } from '@/constants/app';
import {
  FILL,
  GRID_COLORS,
  LINE,
  OPACITY,
  SIZE,
  TOOLS,
} from '@/constants/panels';

const nodes = nodesGenerator(3, 'rectangle');

/**
 * apply different style to nodes to allow for testing every button
 * starting from the first
 */
nodes[0].style.color = 'gray600';
nodes[0].style.size = 'extra-large';
nodes[0].style.line = 'dotted';

const selectedNodeIdsArray = [nodes[0].nodeProps.id, nodes[1].nodeProps.id];
const selectedNodeIds = Object.fromEntries(
  selectedNodeIdsArray.map((id) => [id, true]),
);
const selectedNodes = nodes.filter(
  (node) => node.nodeProps.id in selectedNodeIds,
);

describe('style panel', () => {
  const preloadedState = stateGenerator({
    canvas: { present: { nodes, selectedNodeIds } },
  });

  it('toggles render state based on selectedNodeIds', () => {
    const { rerender } = renderWithProviders(
      <Panels selectedNodeIds={selectedNodeIdsArray} />,
    );

    expect(screen.getByTestId(/style-panel/)).toBeInTheDocument();

    rerender(<Panels selectedNodeIds={[]} />);

    expect(screen.queryByTestId(/style-panel/)).toBeNull();
  });

  describe('colors grid', () => {
    GRID_COLORS.forEach((color) => {
      it(`dispatches nodes update with ${color.name} color`, async () => {
        const { store, user } = renderWithProviders(
          <Panels selectedNodeIds={selectedNodeIdsArray} />,
          { preloadedState },
        );

        await user.click(screen.getByTestId(`${color.value}-color-button`));

        expect(store.dispatch).toHaveBeenCalledWith(
          canvasActions.updateNodes(
            selectedNodes.map((node) => {
              return { ...node, style: { ...node.style, color: color.value } };
            }),
          ),
        );
      });
    });

    it('should have all buttons unselected if selected nodes have different colors', () => {
      renderWithProviders(<Panels selectedNodeIds={selectedNodeIdsArray} />, {
        preloadedState,
      });

      const colorsGrid = screen.getByTestId(/colors-grid/);

      GRID_COLORS.forEach((color) => {
        const colorButton = within(colorsGrid).getByTestId(
          `${color.value}-color-button`,
        );

        expect(colorButton).toHaveAttribute('data-state', 'unchecked');
        expect(colorButton).toHaveAttribute('aria-checked', 'false');
      });
    });
  });

  describe('opacity', () => {
    it('dispatches nodes update with new opacity value', async () => {
      const { user, store } = renderWithProviders(
        <Panels selectedNodeIds={selectedNodeIdsArray} />,
        {
          preloadedState,
        },
      );

      screen.getByTestId(/opacity-slider/).focus();

      await user.keyboard('[ArrowLeft]');

      expect(store.dispatch).toHaveBeenCalledWith(
        canvasActions.updateNodes(
          selectedNodes.map((node) => {
            return {
              ...node,
              style: { ...node.style, opacity: 1 - OPACITY.step },
            };
          }),
        ),
      );
    });
  });

  describe('size', () => {
    SIZE.forEach((size) => {
      it(`dispatches nodes update with ${size.name} size`, async () => {
        const { store, user } = renderWithProviders(
          <Panels selectedNodeIds={selectedNodeIdsArray} />,
          { preloadedState },
        );

        await user.click(screen.getByTestId(`${size.value}-size-button`));

        expect(store.dispatch).toHaveBeenCalledWith(
          canvasActions.updateNodes(
            selectedNodes.map((node) => {
              return { ...node, style: { ...node.style, size: size.value } };
            }),
          ),
        );
      });
    });
  });

  describe('fill', () => {
    FILL.forEach((fill) => {
      it(`dispatches nodes update with ${fill.name} fill`, async () => {
        const { store, user } = renderWithProviders(
          <Panels selectedNodeIds={selectedNodeIdsArray} />,
          { preloadedState },
        );

        await user.click(screen.getByTestId(`${fill.value}-fill-button`));

        expect(store.dispatch).toHaveBeenCalledWith(
          canvasActions.updateNodes(
            selectedNodes.map((node) => {
              return { ...node, style: { ...node.style, fill: fill.value } };
            }),
          ),
        );
      });
    });
  });

  describe('line', () => {
    LINE.forEach((line) => {
      it(`dispatches nodes update with ${line.name} fill`, async () => {
        const { store, user } = renderWithProviders(
          <Panels selectedNodeIds={selectedNodeIdsArray} />,
          { preloadedState },
        );

        await user.click(screen.getByTestId(`${line.value}-line-button`));

        expect(store.dispatch).toHaveBeenCalledWith(
          canvasActions.updateNodes(
            selectedNodes.map((node) => {
              return { ...node, style: { ...node.style, line: line.value } };
            }),
          ),
        );
      });
    });
  });

  describe('animated', () => {
    it(`dispatches nodes update with new animated value`, async () => {
      const { store, user } = renderWithProviders(
        <Panels selectedNodeIds={selectedNodeIdsArray} />,
        { preloadedState },
      );

      await user.click(screen.getByTestId(/animated-toggle/));

      expect(store.dispatch).toHaveBeenCalledWith(
        canvasActions.updateNodes(
          selectedNodes.map((node) => {
            return { ...node, style: { ...node.style, animated: true } };
          }),
        ),
      );
    });
  });
});

describe('delete nodes button', () => {
  const preloadedState = stateGenerator({
    canvas: { present: { nodes, selectedNodeIds } },
  });

  it('toggles the disabled state base on selectedNodeIds', () => {
    const { rerender } = renderWithProviders(
      <Panels selectedNodeIds={selectedNodeIdsArray} />,
    );

    expect(screen.getByTestId(/delete-nodes-button/)).not.toBeDisabled();

    rerender(<Panels selectedNodeIds={[]} />);

    expect(screen.getByTestId(/delete-nodes-button/)).toBeDisabled();
  });

  it('dispatches deleteNodes with correct payload on click', async () => {
    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={selectedNodeIdsArray} />,
      {
        preloadedState,
      },
    );

    await user.click(screen.getByTestId(/delete-nodes-button/));

    expect(store.dispatch).toHaveBeenCalledWith(
      canvasActions.deleteNodes(selectedNodeIdsArray),
    );
  });
});

describe('history buttons', () => {
  it('disables undo history button when there is no past history', () => {
    const preloadedState = stateGenerator({ canvas: { past: [] } });

    renderWithProviders(<Panels selectedNodeIds={[]} />, { preloadedState });

    expect(screen.getByTestId(/undo-history-button/)).toBeDisabled();
  });

  it('disables redo history button when there is no future history', () => {
    const preloadedState = stateGenerator({ canvas: { future: [] } });

    renderWithProviders(<Panels selectedNodeIds={[]} />, { preloadedState });

    expect(screen.getByTestId(/redo-history-button/)).toBeDisabled();
  });

  it('dispatches undo history', async () => {
    const preloadedState = stateGenerator({
      canvas: { past: [{ ...initialCanvasState, nodes }] },
    });

    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={[]} />,
      {
        preloadedState,
      },
    );

    await user.click(screen.getByTestId(/undo-history-button/));

    expect(store.dispatch).toHaveBeenCalledWith(historyActions.undo());
  });

  it('dispatches redo history', async () => {
    const preloadedState = stateGenerator({
      canvas: { future: [{ ...initialCanvasState, nodes }] },
    });

    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={[]} />,
      {
        preloadedState,
      },
    );

    await user.click(screen.getByTestId(/redo-history-button/));

    expect(store.dispatch).toHaveBeenCalledWith(historyActions.redo());
  });
});

describe('zoom panel', () => {
  it('disables zoom out when scale has reached min value', async () => {
    const preloadedState = stateGenerator({
      canvas: { present: { stageConfig: { scale: ZOOM_RANGE[0] } } },
    });

    renderWithProviders(<Panels selectedNodeIds={[]} />, { preloadedState });

    expect(screen.getByTestId(/zoom-out-button/)).toBeDisabled();
  });

  it('disables zoom in when scale has reached max value', async () => {
    const preloadedState = stateGenerator({
      canvas: { present: { stageConfig: { scale: ZOOM_RANGE[1] } } },
    });

    renderWithProviders(<Panels selectedNodeIds={[]} />, { preloadedState });

    expect(screen.getByTestId(/zoom-in-button/)).toBeDisabled();
  });

  it('disables zoom reset when scale is the default value', async () => {
    const preloadedState = stateGenerator({
      canvas: { present: { stageConfig: { scale: DEFAULT_ZOOM_VALUE } } },
    });

    renderWithProviders(<Panels selectedNodeIds={[]} />, { preloadedState });

    expect(screen.getByTestId(/zoom-reset-button/)).toBeDisabled();
  });

  it('zooms in', async () => {
    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={[]} />,
    );

    await user.click(screen.getByTestId(/zoom-in-button/));

    const state = store.getState().canvas.present;

    expect(state.stageConfig.scale).toBeGreaterThan(1);
  });

  it('zooms out', async () => {
    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={[]} />,
    );

    await user.click(screen.getByTestId(/zoom-out-button/));

    const state = store.getState().canvas.present;

    expect(state.stageConfig.scale).toBeLessThan(1);
  });

  it('resets zoom', async () => {
    const preloadedState = stateGenerator({
      canvas: { present: { stageConfig: { scale: 1.5 } } },
    });

    const { store, user } = renderWithProviders(
      <Panels selectedNodeIds={[]} />,
      {
        preloadedState,
      },
    );

    await user.click(screen.getByTestId(/zoom-reset-button/));

    const state = store.getState().canvas.present;

    expect(state.stageConfig.scale).toBe(1);
  });
});

describe('tools panel', () => {
  TOOLS.forEach((tool) => {
    it(`selects ${tool.value} tool`, async () => {
      const { user, store } = renderWithProviders(
        <Panels selectedNodeIds={[]} />,
      );

      await user.click(screen.getByTestId(`tool-button-${tool.value}`));

      expect(store.dispatch).toHaveBeenCalledWith(
        canvasActions.setToolType(tool.value),
      );
    });
  });
});

describe('when hand tool selected', () => {
  it('should not display: style panel, history buttons, nodes delete button', async () => {
    const preloadedState = stateGenerator({
      canvas: { present: { toolType: 'hand', nodes, selectedNodeIds } },
    });

    renderWithProviders(<Panels selectedNodeIds={selectedNodeIdsArray} />, {
      preloadedState,
    });

    expect(screen.queryByTestId(/style-panel/)).toBeNull();
    expect(screen.queryByTestId(/undo-history-button/)).toBeNull();
    expect(screen.queryByTestId(/redo-history-button/)).toBeNull();
    expect(screen.queryByTestId(/delete-nodes-button/)).toBeNull();
  });
});
