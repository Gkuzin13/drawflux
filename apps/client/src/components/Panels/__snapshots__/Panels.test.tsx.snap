// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`style panel > arrow heads > displays correct values 1`] = `
<button
  aria-controls="radix-:rr7:"
  aria-expanded="true"
  aria-haspopup="dialog"
  class="c-dzAnxj c-dzAnxj-eqRZFF-color-secondary c-dzAnxj-bICGYT-align-center c-dzAnxj-hptCGR-size-xs c-dzAnxj-zCRnq-cv"
  data-state="open"
  data-testid="arrow-start-head-trigger"
  title="Arrow Heads — Start"
  type="button"
  value="start"
>
  <svg
    data-testid="minus-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1.75"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <path
      d="M5 12l14 0"
    />
  </svg>
</button>
`;

exports[`style panel > arrow heads > displays correct values 2`] = `
<button
  aria-controls="radix-:rr8:"
  aria-expanded="true"
  aria-haspopup="dialog"
  class="c-dzAnxj c-dzAnxj-eqRZFF-color-secondary c-dzAnxj-bICGYT-align-center c-dzAnxj-hptCGR-size-xs c-dzAnxj-zCRnq-cv"
  data-state="open"
  data-testid="arrow-end-head-trigger"
  title="Arrow Heads — End"
  type="button"
  value="end"
>
  <svg
    data-testid="arrowNarrowRight-icon"
    fill="none"
    height="18"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1.75"
    viewBox="0 0 24 24"
    width="18"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <path
      d="M5 12l14 0"
    />
    <path
      d="M15 16l4 -4"
    />
    <path
      d="M15 8l4 4"
    />
  </svg>
</button>
`;
